Index methods unimplemented in pandas raise RecursionError

pandas.Index.sum(), count(), and other methods not implemented in pandas.Index hang for tens of seconds and then eventually raise RecursionError.

```python

_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/home/crusaderky-nocrypt/miniforge3/envs/dask-expr/lib/python3.12/site-packages/dask/utils.py:1110: in __call__
    return getattr(__obj, self.method)(*args, **kwargs)
../_collection.py:549: in sum
    return new_collection(self.expr.sum(skipna, numeric_only, min_count))
../_collection.py:1279: in new_collection
    meta = expr._meta
/home/crusaderky-nocrypt/miniforge3/envs/dask-expr/lib/python3.12/functools.py:995: in __get__
    val = self.func(instance)
../_reductions.py:381: in _meta
    meta = self._meta_chunk
../_expr.py:176: in __getattr__
    if is_dataframe_like(self._meta) and key in self._meta.columns:
/home/crusaderky-nocrypt/miniforge3/envs/dask-expr/lib/python3.12/functools.py:995: in __get__
    val = self.func(instance)
E   RecursionError: maximum recursion depth exceeded
!!! Recursion detected (same locals & position)
```

# Expected Behaviour
Either implement them or raise AttributeError, like in pandas.
In dask/dask, some of these methods (count) are implemented, while others (sum) raise AttributeError upon compute.

--- a/dask_expr/_collection.py
+++ b/dask_expr/_collection.py
@@ -83,6 +83,7 @@ from dask_expr._repartition import Repartition, RepartitionFreq
 from dask_expr._shuffle import SetIndex, SetIndexBlockwise, SortValues
 from dask_expr._str_accessor import StringAccessor
 from dask_expr._util import (
+    RaiseAttributeError,
     _BackendData,
     _convert_to_list,
     _maybe_from_pandas,
@@ -1886,12 +1887,6 @@ class Index(Series):
         "rename_categories",
     }
 
-    _monotonic_attributes = {
-        "is_monotonic",
-        "is_monotonic_increasing",
-        "is_monotonic_decreasing",
-    }
-
     def __getattr__(self, key):
         if (
             isinstance(self._meta.dtype, pd.CategoricalDtype)
@@ -1900,8 +1895,13 @@ class Index(Series):
             return getattr(self.cat, key)
         elif key in self._dt_attributes:
             return getattr(self.dt, key)
-        elif key in self._monotonic_attributes:
-            return getattr(self, key)
+
+        if hasattr(super(), key):  # Doesn't trigger super().__getattr__
+            # Not a magic attribute. This is a real method or property of Series that
+            # has been overridden by RaiseAttributeError().
+            raise AttributeError(
+                f"{self.__class__.__name__!r} object has no attribute {key!r}"
+            )
         return super().__getattr__(key)
 
     def __repr__(self):
@@ -1920,12 +1920,6 @@ class Index(Series):
     def memory_usage(self, deep=False):
         return new_collection(MemoryUsageIndex(self, deep=deep))
 
-    @property
-    def index(self):
-        raise NotImplementedError(
-            f"{self.__class__.__name__!r} object has no attribute 'index'"
-        )
-
     def shift(self, periods=1, freq=None):
         return new_collection(expr.ShiftIndex(self, periods, freq))
 
@@ -1937,6 +1931,17 @@ class Index(Series):
         if isinstance(self.dtype, pd.CategoricalDtype):
             o.update(self._cat_attributes)
         return list(o)
+
+    # Methods and properties of Series that are not implemented on Index
+    index = RaiseAttributeError()
+    sum = RaiseAttributeError()
+    prod = RaiseAttributeError()
+    count = RaiseAttributeError()
+    mean = RaiseAttributeError()
+    std = RaiseAttributeError()
+    var = RaiseAttributeError()
+    idxmin = RaiseAttributeError()
+    idxmax = RaiseAttributeError()
 
 
 class Scalar(FrameBase):

--- a/dask_expr/_util.py
+++ b/dask_expr/_util.py
@@ -4,7 +4,7 @@ import functools
 from collections import OrderedDict, UserDict
 from collections.abc import Hashable, Sequence
 from types import LambdaType
-from typing import Any, Literal, TypeVar, cast
+from typing import Any, Literal, NoReturn, TypeVar, cast
 
 import dask
 import pandas as pd
@@ -182,3 +182,26 @@ def _maybe_from_pandas(dfs):
         for df in dfs
     ]
     return dfs
+
+
+class RaiseAttributeError:
+    """Method or property defined on superclass, but not on subclass.
+
+    Usage::
+
+        class A:
+            def x(self): ...
+
+        class B(A):
+            x = RaiseAttributeError()
+    """
+
+    name: str
+
+    def __set_name__(self, owner: type, name: str) -> None:
+        self.name = name
+
+    def __get__(self, instance: object | None, owner: type) -> NoReturn:
+        raise AttributeError(
+            f"{owner.__name__!r} object has no attribute {self.name!r}"
+        )

--- a/dask_expr/tests/test_collection.py
+++ b/dask_expr/tests/test_collection.py
@@ -200,7 +200,7 @@ def test_dask(pdf, df):
         ),
         pytest.param(
             lambda df: df.size,
-            marks=pytest.mark.skip(reason="scalars don't work yet"),
+            marks=pytest.mark.xfail(reason="scalars don't work yet"),
         ),
     ],
 )
@@ -214,6 +214,44 @@ def test_reductions(func, pdf, df):
     # check_dtype False because sub-selection of columns that is pushed through
     # is not reflected in the meta calculation
     assert_eq(func(df)["x"], func(pdf)["x"], check_dtype=False)
+
+
+@pytest.mark.parametrize(
+    "func",
+    [
+        M.max,
+        M.min,
+        M.any,
+        M.all,
+        lambda idx: idx.size,
+    ],
+)
+def test_index_reductions(func, pdf, df):
+    result = func(df.index)
+    assert not result.known_divisions
+    assert_eq(result, func(pdf.index))
+
+
+@pytest.mark.parametrize(
+    "func",
+    [
+        lambda idx: idx.index,
+        M.sum,
+        M.prod,
+        M.count,
+        M.mean,
+        M.std,
+        M.var,
+        M.idxmin,
+        M.idxmax,
+    ],
+)
+def test_unimplemented_on_index(func, pdf, df):
+    # Methods/properties of Series that don't exist on Index
+    with pytest.raises(AttributeError):
+        func(pdf.index)
+    with pytest.raises(AttributeError, match="'Index' object has no attribute '"):
+        func(df.index)
 
 
 def test_reduction_on_empty_df():
@@ -2033,11 +2071,6 @@ def test_items(df, pdf):
     for (expect_name, expect_col), (actual_name, actual_col) in zip(expect, actual):
         assert expect_name == actual_name
         assert_eq(expect_col, actual_col)
-
-
-def test_index_index(df):
-    with pytest.raises(NotImplementedError, match="has no"):
-        df.index.index
 
 
 def test_axes(df, pdf):

--- a/dask_expr/tests/test_util.py
+++ b/dask_expr/tests/test_util.py
@@ -1,0 +1,21 @@
+import pytest
+
+from dask_expr._util import RaiseAttributeError
+
+
+def test_raises_attribute_error():
+    class A:
+        def x(self):
+            ...
+
+    class B(A):
+        x = RaiseAttributeError()
+
+    assert hasattr(A, "x")
+    assert hasattr(A(), "x")
+    assert not hasattr(B, "x")
+    assert not hasattr(B(), "x")
+    with pytest.raises(AttributeError, match="'B' object has no attribute 'x'"):
+        B.x
+    with pytest.raises(AttributeError, match="'B' object has no attribute 'x'"):
+        B().x

